### PR TITLE
Dm salmon

### PR DESCRIPTION
**Pull request by ieguinoa**
Data manager for Salmon, code is mainly based on the Kallisto index builder dm.

### CODE REVIEW

**Comment by lparsons** [comment]
Is there a reason this hasn't been merged? Looks like a really useful tool.

File: data_managers/data_manager_salmon_index_builder/tool_data_table_conf.xml.sample [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<tables>
+    <!-- Locations of indexes in salmon mapper format -->
+    <table name="salmon_indexes" comment_char="#" allow_duplicate_entries="False">
```
Comment (bgruening) [inline data_managers/data_manager_salmon_index_builder/tool_data_table_conf.xml.sample]: I guess either the data table should be versioned or column should be added that contains the version. As we could not find a consensus during GCC, both is fine and you should pick one :)

File: data_managers/data_manager_salmon_index_builder/tool_data_table_conf.xml.sample [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+        <columns>value, dbkey, name, path</columns>
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/tool_data_table_conf.xml.sample]:
```suggestion
        <columns>value, dbkey, name, path, version</columns>
```

File: data_managers/data_manager_salmon_index_builder/tool_data_table_conf.xml.sample [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+        <file path="tool-data/salmon_indexes.loc" />
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/tool_data_table_conf.xml.sample]:
```suggestion
        <file path="tool-data/salmon_indexes_versioned.loc" />
```

File: data_managers/data_manager_salmon_index_builder/data_manager_conf.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+                <column name="dbkey" />
+                <column name="name" />
+                <column name="path" output_ref="out_file" >
+                    <move type="directory" relativize_symlinks="True">
+                        <!-- <source>${path}</source>--> <!-- out_file.extra_files_path is used as base by default --> <!-- if no source, eg for type=directory, then refers to base -->
+                        <target base="${GALAXY_DATA_MANAGER_DATA_PATH}">${dbkey}/salmon_index/${value}</target>
+                    </move>
+                    <value_translation>${GALAXY_DATA_MANAGER_DATA_PATH}/${dbkey}/salmon_index/${value}/${path}</value_translation>
+                    <value_translation type="function">abspath</value_translation>
+                </column>
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/data_manager_conf.xml]:
```suggestion
                </column>
                <column name="version" />
```

File: data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+# Based heavily on the kallisto data manager wrapper script by iuc
+from __future__ import print_function
+
+import argparse
+import os
+import subprocess
+import sys
+from json import dumps, loads
+
+DEFAULT_DATA_TABLE_NAME = "salmon_indexes"
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.py]:
```suggestion
DEFAULT_DATA_TABLE_NAME = "salmon_indexes_versioned"
```

File: data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+def main():
+    # Parse Command Line
+    parser = argparse.ArgumentParser()
+    parser.add_argument( '--output', dest='output', action='store', type=str, default=None )
+    parser.add_argument( '--fasta_filename', dest='fasta_filename', action='store', type=str, default=None )
+    parser.add_argument( '--fasta_dbkey', dest='fasta_dbkey', action='store', type=str, default=None )
+    parser.add_argument( '--fasta_description', dest='fasta_description', action='store', type=str, default=None )
+    parser.add_argument( '--data_table_name', dest='data_table_name', action='store', type=str, default='salmon_indexes' )
+    parser.add_argument( '-k', '--kmer_size', dest='kmer_size', action='store', type=str, help='kmer_size' )
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.py]:
```suggestion
    parser.add_argument( '-k', '--kmer_size', dest='kmer_size', action='store', type=str, help='kmer_size' )
    parser.add_argument( '--index_version' )
```

File: data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    if options.kmer_size != '':
+        args.append('-k')
+        args.append(options.kmer_size)
+    args.extend( [ '-t', options.fasta_filename, '-i', target_directory ] )
+    proc = subprocess.Popen( args=args, shell=False)
+    return_code = proc.wait()
+    if return_code:
+        print("Error building index.", file=sys.stderr)
+        sys.exit( return_code )
+    data_table_entry = dict( value=sequence_id, dbkey=options.fasta_dbkey, name=sequence_name, path=sequence_id )
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.py]:
```suggestion
    data_table_entry = dict( value=sequence_id, dbkey=options.fasta_dbkey, name=sequence_name, path=sequence_id, version=options.index_version )
```

File: data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.xml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+<tool id="salmon_index_builder_data_manager" name="Salmon" tool_type="manage_data" version="0.9.1">
+    <description>index builder</description>
+    <requirements>
+        <requirement type="package" version="0.9.1">salmon</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.xml]:
```suggestion
   <macros>
       <token name="@IDX_VERSION@">0</token>
   </macros>
    <command detect_errors="exit_code"><![CDATA[
```

File: data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.xml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        python '$__tool_directory__/salmon_index_builder.py' --output '${out_file}'
+            --fasta_filename '${all_fasta_source.fields.path}'
+            --fasta_dbkey '${all_fasta_source.fields.dbkey}'
+            --fasta_description '${all_fasta_source.fields.name}'
+            --kmer_size "${kmer_size}"
+            --data_table_name salmon_indexes
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/data_manager/salmon_index_builder.xml]:
```suggestion
            --data_table_name salmon_indexes
            --index_version @IDX_VERSION@
```

File: data_managers/data_manager_salmon_index_builder/tool-data/salmon_indexes.loc.sample [hunk 10]
```diff
@@ -0,0 +1,28 @@
+# salmon_indexes.loc.sample
```
Comment (wm75) [inline data_managers/data_manager_salmon_index_builder/tool-data/salmon_indexes.loc.sample]: for a versioned data table, this file would have to be renamed and the additional column be documented

**Comment by wm75** [comment]
Oh and if you're looking for further inspiration, you could have a look at #2489, which I just opened today, and see if you like the idea. You could pick up the general refactoring concept if you like it (but keep passing command line arguments instead of a yaml file for now).

**Comment by ieguinoa** [comment]
This version of the DM creates indexes of version q5 using the tool version 0.11.2 
I tracked down the versions of indexes created by Salmon that can be found in header.json file in the index directory. It seems to be that:

index version | works with tool version(s)
q1|0.6
q3|0.7
q4|0.8
q5|[0.9 - 0.13]
q6|0.14 

so, when salmon wrappers gets an update to current version (0.14) then the index version also needs to be updated to use q6.

Ignacio

**Comment by bgruening** [comment]
Merci!

**Pull request merged**